src/agents/OIDCCredentialAgent.js

```diff
@@ -23,29 +26,57 @@ export default class OIDCCredentialAgent {
     return this._accessToken;
   }
 
-  getCredentials() {
+  async getCredentials() {
     if (this.credentialsPromise && this.credentialsExpire > new Date()) {
       return this.credentialsPromise;
     }
 
     // Call the oidcCredentials endpoint with the access token.
     const loginBaseUrl = 'https://login.taskcluster.net';
     const url = `${loginBaseUrl}/v1/oidc-credentials/${this.oidcProvider}`;
+    const mutex = new Mutex();
+
+    await mutex.lock(MUTEX_LOCK_KEY);
+
+    const session = localStorage.getItem(MUTEX_LOCK_KEY);
+
+    // If the session exists, it was set by the leader mutex. Grab the
+    // credentials from storage and pass them on.
+    if (session) {
+      const response = JSON.parse(session);
 
-    this.credentialsPromise = fetch(url, {
-      headers: {
-        Authorization: `Bearer ${this.accessToken}`
-      }
-    })
-    .then((response) => {
       this.credentialsExpire = new Date(response.expires);
+      this.credentialsPromise = Promise.resolve(response.credentials);
+    } else {
+      // With nothing in localStorage, we are the leader mutex, and will
+      // be logging in for all other sessions. Make the request, set the
+      // value in storage, then unlock the mutex for peers.
+      // We will also blow away the session from storage in 30 seconds
+      // as a recovery mechanism, and to also ensure that the next login
+      // gets a clean slate.
+      this.credentialsPromise = new Promise(async (resolve, reject) => {
+        try {
+          const response = await fetch(url, {
+            headers: {
+              Authorization: `Bearer ${this.accessToken}`
+            }
+          });
 
-      return response.credentials;
-    })
-    .catch((err) => {
-      this.credentialsPromise = null;
-      throw err;
-    });
+          this.credentialsExpire = new Date(response.expires);
+          localStorage.setItem(MUTEX_LOCK_KEY, JSON.stringify(response));
+
+          setTimeout(() => {
+            localStorage.removeItem(MUTEX_LOCK_KEY);
+          }, 30000);
+
+          await mutex.release(MUTEX_LOCK_KEY);
+          resolve(response.credentials);
+        } catch (err) {
+          this.credentialsPromise = null;
+          reject(err);
```

Comment:
Sort of separate from this bug, but it would be nice to have some logging of this error to console, as well as some kind of event emitted that the tools site could use to alert the user.  I keep finding myself logged out of tools with nothing in the console, and often I don't notice until I've opened and closed a few tabs so even if there was data in the console, I would have missed it.  A little doorhanger saying "You were logged out. _Details_" would help debug.

Reply:
That would have to exist in the consumer side; we reject here and it's up to the consumer to log or handle afterwards.